### PR TITLE
Explore: Implement caching of logs result from navigation

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
The next PR from **Logs navigation** series. This ones implements caching of the last 10 results run from navigation. The scope of this PR is only caching and functionality, not UI (more UI changes coming).

1. I had to update logsModel to store queries it was run with. This way we can reliable reset navigation and logsResultCache when the queries that we run change. Otherwise, we could use only absoluteRange. Previously we used queries object, but this was buggy as it did reset navigation and cache when you just edited query (without running it). 
2. I am caching last 5 sets of results. We can do in the future/polishing phase some benchmarking of how many results should we save. 

3. I still need to finish writing tests, but I think it is ready for functionality/code review and testing. 

https://user-images.githubusercontent.com/30407135/117171563-43a84680-adcb-11eb-8bbf-190cf565ed02.mov

https://user-images.githubusercontent.com/30407135/117173020-a3ebb800-adcc-11eb-8d51-e5e9116b6808.mov

### CODE REVIEW

**Comment by ifrost** [comment]
When reaching no logs / no data range all previous ranges disappear from the right-side navigation:

| master | branch |
|--------|--------|
| logs-1 | logs-2 |

**Comment by ivanahuckova** [comment]
>When reaching no logs / no data range all previous ranges disappear from the right-side navigation: 

That's not good! I'll look at it. Nice catch.

File: public/app/features/explore/LogsContainer.tsx [hunk 1]
```diff
@@ -28,10 +27,78 @@ interface LogsContainerProps {
...
+
+  componentDidUpdate(prevProps: LogsContainerProps) {
+    const { logsResult, absoluteRange } = this.props;
+    // If new results, update cache
+    if (logsResult && !isEqual(logsResult, prevProps.logsResult)) {
+      // If queries were changed, reset cache and start fresh
+      if (!isEqual(logsResult.queries, prevProps.logsResult?.queries)) {
+        this.logRowsCache.reset();
+      }
+      // Update state and add to logResults and absolutRange to chache
```
Comment (ifrost) [inline public/app/features/explore/LogsContainer.tsx:61]: I think it's spelled "cha-cha-cha" :D 

```suggestion
      // Update state and add to logResults and absolutRange to cache
```

chachacha

Comment (ivanahuckova) [reply public/app/features/explore/LogsContainer.tsx:61]: 🤣🤣🤣🤣

**Comment by ivanahuckova** [comment]
> When reaching no logs / no data range all previous ranges disappear from the right-side navigation:

Again nice catch! Here is the fix: https://github.com/grafana/grafana/pull/33744/commits/afb77f1d34c20164be598b123ad3843495fd56ed.

**Comment by ifrost** [comment]
🤔 I think the behaviour of Inspect Data tab has changed slightly.

Model's `series` prop ([I think this one](https://github.com/grafana/grafana/blob/main/public/app/core/logs_model.ts#L220)) is updated only when data is loaded.

Without caching: data for the selected range was always loaded and `series` property was updated. 
With caching: the property contains only last loaded data and doesn't change when the range is picked from the cache.

I think the same property is used by Inspect Data tab, meaning it always shows data for last loaded range, not last selected range.

**Pull request closed without merging**